src/test/java/com/google/devtools/build/lib/skylark/SkylarkIntegrationTest.java

```diff
@@ -3160,6 +3160,52 @@ public void testSplitEmptySeparator() throws Exception {
     getConfiguredTarget("//test:r");
   }
 
+  @Test
+  public void testHashFrozenList() throws Exception {
+    setSkylarkSemanticsOptions("--incompatible_disallow_hashing_frozen_mutables=false");
+    scratch.file("test/extension.bzl", "y = []");
+
+    scratch.file(
+        "test/BUILD",
+        "load('//test:extension.bzl', 'y')",
+        "{y: 1}",
+        "cc_library(name = 'r')");
+
+    getConfiguredTarget("//test:r");
+  }
+
+  @Test
+  public void testHashFrozenListForbidden() throws Exception {
+    setSkylarkSemanticsOptions("--incompatible_disallow_hashing_frozen_mutables=true");
+    scratch.file("test/extension.bzl", "y = []");
+
+    scratch.file(
+        "test/BUILD",
+        "load('//test:extension.bzl', 'y')",
+        "{y: 1}",
+        "cc_library(name = 'r')");
+
+    reporter.removeHandler(failFastHandler);
+    getConfiguredTarget("//test:r");
+    assertContainsEvent("unhashable type: 'list'");
+  }
+
+  @Test
+  public void testHashFrozenDeepMutableForbidden() throws Exception {
+    setSkylarkSemanticsOptions("--incompatible_disallow_hashing_frozen_mutables=true");
+    scratch.file("test/extension.bzl", "y = {}");
+
+    scratch.file(
+        "test/BUILD",
+        "load('//test:extension.bzl', 'y')",
+        "{('a', (y,), True): None}",
+        "cc_library(name = 'r')");
+
+    reporter.removeHandler(failFastHandler);
+    getConfiguredTarget("//test:r");
+    assertContainsEvent("unhashable type: 'tuple'");
```

Comment:
This error message is a bit confusing, implying that tuples are never hashable. Would it be feasible to change it to mention the actual unhashable type within this PR? The similar error in Python is "unhashable type: 'dict'".

Reply:
We might want to submit this PR first (and hopefully be included in Bazel 0.29) and iterate later.

Reply:
Since this is really the object that is unhashable rather than the type, how about changing the message to
"unhashable object of type: dict"

Reply:
btw the same happens, regardless of this PR, if you try something like `{([],): None}` you will also get the same error `unhashable type: 'tuple'`.
I could still fix it in this PR by modifying the [SkylarkValue](https://github.com/bazelbuild/bazel/blob/master/src/main/java/com/google/devtools/build/lib/skylarkinterface/SkylarkValue.java) interface to get the mutable/hashable object type within an immutable structure (like tuple), but I'm not sure if that's the best way to do so.

Reply:
I like @aiuto's suggestion.